sklearn/neighbors/_base.py

```diff
@@ -553,12 +567,29 @@ def _fit(self, X, y=None):
                 **self.effective_metric_params_,
             )
         elif self._fit_method == "kd_tree":
-            self._tree = KDTree(
-                X,
-                self.leaf_size,
-                metric=self.effective_metric_,
-                **self.effective_metric_params_,
-            )
+            if (
+                self.effective_metric_ == "minkowski"
+                and "w" in self.effective_metric_params_
+            ):
+                warnings.warn(
+                    "KDTree does not support Weighted Minkowski. "
+                    "Falling back on BallTree."
+                )
+                self._fit_method = "ball_tree"
+                self._tree = BallTree(
+                    X,
+                    self.leaf_size,
+                    metric=self.effective_metric_,
+                    **self.effective_metric_params_,
+                )
+
+            else:
+                self._tree = KDTree(
+                    X,
+                    self.leaf_size,
+                    metric=self.effective_metric_,
+                    **self.effective_metric_params_,
+                )
```

Comment:
If the user asks for `algorithm="kd_tree"` explicitly (not "auto"),  I would let the invalid metric exception be raised  as is currently done with the deprecated wminknowski:


```python
In [5]: from sklearn.neighbors import NearestNeighbors

In [6]: NearestNeighbors(metric="wminknowski", metric_params={"w": [0.1, 1.]}).fit([[1, 0], [0, 1]]).kneighbors([[0, 1.2]])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [6], in <module>
----> 1 NearestNeighbors(metric="wminknowski", metric_params={"w": [0.1, 1.]}).fit([[1, 0], [0, 1]]).kneighbors([[0, 1.2]])

File ~/code/scikit-learn/sklearn/neighbors/_unsupervised.py:168, in NearestNeighbors.fit(self, X, y)
    151 def fit(self, X, y=None):
    152     """Fit the nearest neighbors estimator from the training dataset.
    153 
    154     Parameters
   (...)
    166         The fitted nearest neighbors estimator.
    167     """
--> 168     return self._fit(X)

File ~/code/scikit-learn/sklearn/neighbors/_base.py:444, in NeighborsBase._fit(self, X, y)
    441     if not isinstance(X, (KDTree, BallTree, NeighborsBase)):
    442         X = self._validate_data(X, accept_sparse="csr", order="C")
--> 444 self._check_algorithm_metric()
    445 if self.metric_params is None:
    446     self.effective_metric_params_ = {}

File ~/code/scikit-learn/sklearn/neighbors/_base.py:379, in NeighborsBase._check_algorithm_metric(self)
    372         raise ValueError(
    373             "kd_tree does not support callable metric '%s'"
    374             "Function call overhead will result"
    375             "in very poor performance."
    376             % self._metric
    377         )
    378 elif self._metric not in VALID_METRICS[alg_check]:
--> 379     raise ValueError(
    380         "Metric '%s' not valid. Use "
    381         "sorted(sklearn.neighbors.VALID_METRICS['%s']) "
    382         "to get valid options. "
    383         "Metric can also be a callable function." % (self._metric, alg_check)
    384     )
    386 if self.metric_params is not None and "p" in self.metric_params:
    387     if self.p is not None:

ValueError: Metric 'wminknowski' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.
```

The automatic switch should only happen when `algorithm="auto"`.

Reply:
But indeed we might need to special-case the error message for this beast:

```python
raise ValueError("algorithm='kd_tree' is not valid for metric='minkowski' with a weight parameter 'w': try algorithm='ball_tree' or algorithm='brute' instead.")
```

or something similar.